File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map <char, int> mp;
        int start=0,end=0,len=0,ml=0;
        for(int i=0;i<s.length();i++){
            char ch=s[i];
            if(mp.find(ch)!=mp.end()  && mp[ch] >= start){
                start=mp[ch]+1;
            }
            else{
                len=i-start+1;
                ml=max(ml,len);
            }
            mp[ch]=i;
        }
        return ml;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map <Character, Integer> map= new HashMap<>();
        int start=0, end=0;
        int maxl=0;
        for(int i=0;i<s.length();i++){
            if(map.containsKey(s.charAt(i)) && map.get(s.charAt(i)) >= start) {
                start = map.get(s.charAt(i)) + 1;
            }
            map.put(s.charAt(i),i);
            int len=i-start+1;
            if(len>maxl) maxl=len;
            
        }
        return maxl;
    }
}